tools/java/org/unicode/cldr/util/XMLNormalizingLoader.java

```diff
@@ -0,0 +1,686 @@
+            this.localeId = localeId;
+            this.dirs = dirs;
+            this.minimalDraftStatus = minimalDraftStatus;
+            this.hashCode = Objects.hash(this.localeId, this.dirs, this.minimalDraftStatus);
+        }
+
+        @Override
+        public int hashCode() {
+            return hashCode;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj == null) {
+                return false;
+            }
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+            XMLSourceCacheKey other = (XMLSourceCacheKey) obj;
+            if (!Objects.equals(dirs, other.dirs)) {
+                return false;
+            }
+            if (minimalDraftStatus != other.minimalDraftStatus) {
+                return false;
+            }
+            if (!Objects.equals(localeId, other.localeId)) {
+                return false;
+            }
+            return true;
+        }
+    }
+
+    public static XMLSource getFrozenInstance(String localeId, List<File> dirs, DraftStatus minimalDraftStatus) {
+        XMLSourceCacheKey key = new XMLSourceCacheKey(localeId, dirs, minimalDraftStatus);
+        return cache.getUnchecked(key);
+    }
+
+    private static XMLSource makeXMLSource(XMLSourceCacheKey key) {
+        if (key.dirs == null || key.dirs.size() == 0) {
+            return null;
+        }
+
+        XMLSource source = null;
+        if (key.dirs.size() == 1) {
+            File file = new File(key.dirs.get(0), key.localeId + ".xml");
+            source = loadXMLFile(file, key.localeId, key.minimalDraftStatus);
+            source.freeze();
+            return source;
+        }
+
+        // if contains more than one file, make XMLSource from each file and then combine them to a combined XMLSource,
+        // so that can cache single file XMLSource as well as combined XMLSource
+        List<XMLSource> list = new ArrayList<>();
+        List<File> dirList = new ArrayList<>();
+        for (File dir: key.dirs) {
+            dirList.clear();
+            dirList.add(dir);
+            XMLSourceCacheKey singleKey = new XMLSourceCacheKey(key.localeId, dirList, key.minimalDraftStatus);
+            XMLSource singleSource = cache.getUnchecked(singleKey);
+            list.add(singleSource);
+        }
+
+        source = list.get(0).cloneAsThawed();
+        for (int i = 1; i < list.size(); i++) {
+            XMLSource other = list.get(i);
+            source.putAll(other, 0); // 0 --> merge_keep_mine
+            source.getXpathComments().joinAll(other.getXpathComments());
+        }
+        source.freeze();
+        return source;
+    }
+
+    public static XMLSource loadXMLFile(File f, String localeId, DraftStatus minimalDraftStatus) {
+        String fullFileName = null;
```

Comment:
Nitpick: you can move `String fullFileName = f.getCanonicalPath();`  to be the first thing in the main `try { }`
It is not used outside the try, and that already catches the exception.
So instead of 2 try-catch blocks you only have one.